cmd/node-agent/main.go

```diff
@@ -0,0 +1,220 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-systemd/dbus"
+	"github.com/coreos/go-systemd/unit"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_3"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	semver "k8s.io/kubernetes/pkg/version"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const (
+	myPodNameEnvName               = "MY_POD_NAME"
+	desiredVersionAnnotation       = "updates.coreos.com/desired-version"
+	currentVersionAnnotation       = "updates.coreos.com/current-version"
+	desiredKubeletConfigAnnotation = "updates.coreos.com/desired-kubelet-config"
+	currentKubeletConfigAnnotation = "updates.coreos.com/current-kubelet-config"
+	kubeletConfigService           = "kubelet-config"
+	kubeletVersionService          = "kubelet-version"
+	kubeletConfigServicePath       = "/etc/systemd/system/kubelet.service.d/50-kubelet-config.conf"
+	kubeletVersionServicePath      = "/etc/systemd/system/kubelet.service.d/70-kubelet-version.conf"
+)
+
+var (
+	sysdConn *dbus.Conn
+	client   clientset.Interface
+
+	kubeAPIAddr = fmt.Sprintf("https://%s:%s", os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT_HTTPS"))
+	myPodName   = os.Getenv(myPodNameEnvName)
+)
+
+func main() {
+	if myPodName == "" {
+		glog.Fatalf("%s env var not set.", myPodNameEnvName)
+	}
+	run()
+}
+
+func run() {
+	client = newAPIClient()
+	nodename := getNodeName()
+
+	glog.Infof("starting node agent, watching node: %s", nodename)
+
+	opts := api.ListOptions{
+		FieldSelector: fields.ParseSelectorOrDie(fmt.Sprintf("metadata.name=%s", nodename)),
+	}
+	_, nodeController := framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(lo api.ListOptions) (runtime.Object, error) {
+				return client.Core().Nodes().List(opts)
+			},
+			WatchFunc: func(lo api.ListOptions) (watch.Interface, error) {
+				return client.Core().Nodes().Watch(opts)
+			},
+		},
+		&v1.Node{},
+		30*time.Second,
+		framework.ResourceEventHandlerFuncs{
+			UpdateFunc: nodeUpdateCallback,
+		},
+	)
+
+	var err error
+	sysdConn, err = dbus.New()
+	if err != nil {
+		glog.Fatal(err)
+	}
+
+	for {
+		// We say to never stop, but this method will
+		// return if it fails to contact an API server.
+		// This service should be resilient to temporary
+		// loss of an API server.
```

Comment:
Very true, but I figure that temporary loss of an API server is not an explicit "failure" condition that we need to surface. If it's more than temporary, the cluster admin is likely being paged anyways.